cmd/apprepository-controller/server/controller_test.go

```diff
@@ -1883,6 +1975,59 @@ func TestGenerateJobName(t *testing.T) {
 	}
 }
 
+func TestIntervalToCron(t *testing.T) {
+	testCases := []struct {
+		name         string
+		interval     string
+		expectedCron string
+	}{
+		{
+			name:         "good interval, every 2 nanoseconds",
+			interval:     "2ns",
+			expectedCron: "*/1 * * * *",
+		},
+		{
+			name:         "good interval, every 2 microseconds",
+			interval:     "2us",
+			expectedCron: "*/1 * * * *",
+		},
+		{
+			name:         "good interval, every 2 milliseconds",
+			interval:     "2ms",
+			expectedCron: "*/1 * * * *",
+		},
+		{
+			name:         "good interval, every 2 seconds",
+			interval:     "2s",
+			expectedCron: "*/1 * * * *",
+		},
+		{
+			name:         "good interval, every two minutes",
+			interval:     "2m",
+			expectedCron: "*/2 * * * *",
+		},
+		{
+			name:         "good interval, every two hours",
+			interval:     "2h",
+			expectedCron: "*/120 * * * *",
```

Comment:
Is that actually valid? I thought Cron was stateless (and so effectively reset at the turn of each hour for a minutes range here). I'm not 100%, but worth checking (see some [notes on the wikipedia article](https://en.wikipedia.org/wiki/Cron#Non-standard_characters) - not a source of truth obviously).

Could be worth using https://pkg.go.dev/github.com/robfig/cron#Every , not sure.

Reply:
Thanks for your comments and pointers, I've had to revisit my knowledge about how `cron` works internally. 

The expression I was building is syntactically valid, that's why I thought it was valid (=meaningful). However, it is semantically pointless given how the cron works.
Kubernetes is using the `vixie-cron` implementation, the POSIX-compliant implementation developed for BSD (back in 1988). 

NOTE: There are some other implementations, like `dcron`, `fcron`, `bcron` and `cronie`. The latter, is a `vixie-cron` fork maintained by Fedora where, for instance, [does validate the `step` field](https://github.com/cronie-crond/cronie/blob/master/readme.md) and therefore makes the discussed expression invalid syntactically.

Well, so coming back to the vixie implementation, the step can be any number:

https://github.com/vixie/cron/blob/2e298fc87bf6f9220d09e60162ebf909a607c4fb/entry.c#L518-L521

But, it is going to, silently, drop the invalid values:

https://github.com/vixie/cron/blob/2e298fc87bf6f9220d09e60162ebf909a607c4fb/entry.c#L580-L584

In short, as you pointed out, in an interval `0-59`, the only possible values for step are `[1-59]`.

Unfortunately, this is something that k8s is NOT validating:

````
k get cronjob        
NAME                  SCHEDULE       SUSPEND   ACTIVE   LAST SCHEDULE   AGE
kubernetes-cron-job   */61 * * * *   False     0        <none>          20m
````

> Note the "last schedule" equals to "none"

The library you pointed out seems to be a scheduler itself, as it does not generate the crontab thing, but actually runs a timer to calculate the next executions.

So, it seems we might need to make the function a bit smarter :S and limit the values + perform the proper conversions between time units.

Reply:
> The library you pointed out seems to be a scheduler itself, as it does not generate the crontab thing, but actually runs a timer to calculate the next executions.

Yes, sorry - I realised the library was for a scheduler, but had thought that function I linked to was for converting a duration into a crontab format, but seems it is not.